presentation-compiler/src/main/dotty/tools/pc/completions/CompletionProvider.scala

```diff
@@ -45,23 +50,42 @@ class CompletionProvider(
     val uri = params.uri().nn
     val text = params.text().nn
 
-    val code = applyCompletionCursor(params)
+    val (wasCursorApplied, code) = applyCompletionCursor(params)
     val sourceFile = SourceFile.virtual(uri, code)
+
+    /** Creating a new fresh driver is way slower than reusing existing one,
+     *  but runnig a compilation has side effects that modifies the state of the driver.
+     *  We don't want to affect cachingDriver state with compilation including "CURSOR" suffix.
+     *
+     *  We could in theory save this fresh driver for reuse, but it is a choice between extra memory usage and speed.
+     *  The scenario in which "CURSOR" is applied (empty query or query equal to any keyword) has a slim chance of happening.
+     */
+
+    val driver = if wasCursorApplied then freshDriver() else cachingDriver
```

Comment:
The cursor shouldn't be applied rarely, does it make sense to treat it specially here?

Reply:
We have to treat it specially. Running a compilation mutates the context / changes diagnostics. There is no way to use a Driver without mutating a context, thus to not change and reuse previous compilation, we have to create a new Driver

Reply:
We may try saving a context and replacing it, but I'm not whether this is the best idea. As far as I know, this approach was tried when @tanishiking was adding compilation caching in a very first place. I also think that this is unnecessary overengineering.